gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/MaximumVersion.java

```diff
@@ -0,0 +1,91 @@
+/*
+ * (c) Copyright 2018 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.gradle.dist;
+
+import com.palantir.logsafe.UnsafeArg;
+import com.palantir.logsafe.exceptions.SafeIllegalArgumentException;
+import com.palantir.sls.versions.OrderableSlsVersion;
+import com.palantir.sls.versions.SlsVersionMatcher;
+import com.palantir.sls.versions.SlsVersionType;
+import com.palantir.sls.versions.VersionComparator;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.function.Function;
+import java.util.function.Supplier;
+import java.util.stream.Stream;
+
+/**
+ * A typed version of a maximum version as declared in {@link RecommendedProductDependency#maximumVersion}.
+ * It can be represented by either:
+ * <ul>
+ *     <li>an {@link OrderableSlsVersion}, or</li>
+ *     <li>an {@link SlsVersionMatcher}</li>
+ * </ul>
+ * Hence, this class encodes a functional either-type.
+ * <p>
+ * The dichotomy is necessary because {@link SlsVersionMatcher}, though it has some overlap with
+ * {@link OrderableSlsVersion}, doesn't handle all the cases that the latter supports. For instance, it can't represent
+ * non-release versions (see version types enum {@link SlsVersionType}).
+ */
+abstract class MaximumVersion implements Comparable<MaximumVersion> {
+    /**
+     * Functional abstraction for the visitor pattern.
+     * Handles both possible states of this MaximumVersion (where it is represented by an {@link OrderableSlsVersion}
+     * or an {@link SlsVersionMatcher}), without requiring the caller to know in advance which one it is.
+     */
+    abstract <T> T fold(
```

Comment:
I know this is a cool way to achieve pattern matching-style stuff, but it's pretty confusing to read.  Please could you just use the normal MaximumVersionVisitor thing instead so that it's clear there's no magic here?

Reply:
I don't think this is magic, it's pretty clear what this does coming from a functional programming background. (if the logic were confusing though, that's another issue) 
However it's not a lot more code to write visitors so I can do that.

Reply:
This may look perfectly readable to you, but I think we need to try to make it super super obvious for any future maintainers.  Given that all the conjure union types just use FooVisitor, I think the consistency will be helpful to readers.

Aiming to use the simplest, most boring solution that works here!